onnx/cpp2py_export.cc

```diff
@@ -115,13 +115,187 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
+      .def_property_readonly(
+          "_default_value",
+          [](OpSchema::Attribute* attr) -> py::bytes {
+            std::string out;
+            attr->default_value.SerializeToString(&out);
+            return out;
+          })
+      .def_readonly("required", &OpSchema::Attribute::required);
+
+  py::class_<OpSchema::TypeConstraintParam>(op_schema, "TypeConstraintParam")
+      .def(
+          py::init<std::string, std::vector<std::string>, std::string>(),
+          py::arg("type_param_str"),
+          py::arg("allowed_type_strs"),
+          py::arg("description") = "")
+      .def_readonly("type_param_str", &OpSchema::TypeConstraintParam::type_param_str)
+      .def_readonly("allowed_type_strs", &OpSchema::TypeConstraintParam::allowed_type_strs)
+      .def_readonly("description", &OpSchema::TypeConstraintParam::description);
+
+  py::class_<OpSchema::FormalParameter>(op_schema, "FormalParameter")
+      .def(
+          "__init__",
+          [](OpSchema::FormalParameter& self,
+             std::string name,
+             std::string type_str,
+             const std::string& description,
+             OpSchema::FormalParameterOption param_option,
+             bool is_homogeneous,
+             int min_arity,
+             OpSchema::DifferentiationCategory differentiation_category) {
+            // Use a lambda to swap the order of the arguments to match the Python API
+            new (&self) OpSchema::FormalParameter(
+                name, description, type_str, param_option, is_homogeneous, min_arity, differentiation_category);
+          },
+          py::arg("name"),
+          py::arg("type_str"),
+          py::arg("description") = "",
+          py::arg("param_option") = OpSchema::Single,
+          py::arg("is_homogeneous") = true,
+          py::arg("min_arity") = 1,
+          py::arg("differentiation_category") = OpSchema::DifferentiationCategory::Unknown)
+
+      .def_property_readonly("name", &OpSchema::FormalParameter::GetName)
+      .def_property_readonly("types", &OpSchema::FormalParameter::GetTypes)
+      .def_property_readonly("type_str", &OpSchema::FormalParameter::GetTypeStr)
+      .def_property_readonly("description", &OpSchema::FormalParameter::GetDescription)
+      .def_property_readonly("option", &OpSchema::FormalParameter::GetOption)
+      .def_property_readonly("is_homogeneous", &OpSchema::FormalParameter::GetIsHomogeneous)
+      .def_property_readonly("min_arity", &OpSchema::FormalParameter::GetMinArity)
+      .def_property_readonly("differentiation_category", &OpSchema::FormalParameter::GetDifferentiationCategory)
+      // Legacy camel cased names. We retain them for backward compatibility.
+      .def_property_readonly("typeStr", &OpSchema::FormalParameter::GetTypeStr)
+      .def_property_readonly("isHomogeneous", &OpSchema::FormalParameter::GetIsHomogeneous)
+      .def_property_readonly("differentiationCategory", &OpSchema::FormalParameter::GetDifferentiationCategory);
```

Comment:
> Potentially it may bring a lot of deprecation message.

That's what I worry, but I also think that's OK. I will add a message if there's no objection.

Reply:
Done

Reply:
I wonder what does deprecation mean in that case? Is removed in e.g. 2 releases? or is there again discussed? or the other variant we recommended? Am just curious ;-)

Reply:
I am proposing to remove on 1.16. Tracked by #5074. The recommended variants are the snake_case ones